File: src/main/java/com/baulsupp/oksocial/okhttp/OkHttpResponseExtractor.java
package com.baulsupp.oksocial.okhttp;

import com.baulsupp.oksocial.output.ResponseExtractor;
import com.baulsupp.oksocial.output.util.JsonUtil;
import java.util.List;
import java.util.Optional;
import okhttp3.MediaType;
import okhttp3.Response;
import okio.BufferedSource;

import static java.util.Optional.of;

public class OkHttpResponseExtractor implements ResponseExtractor<Response> {
  @Override public Optional<String> mimeType(Response response) {
    if (response.body() == null) {
      return Optional.empty();
    }

    String host = response.request().url().host();
    MediaType mediaType = response.body().contentType();

    if (mediaType == null) {
      return Optional.empty();
    }

    if (host.equals("graph.facebook.com") && mediaType.subtype().equals("javascript")) {
      return of(JsonUtil.JSON);
    }

    return of(mediaType.toString());
  }

  @Override public BufferedSource source(Response response) {
    return response.body().source();
  }

  @Override public String filename(Response response) {
    List<String> segments = response.request().url().pathSegments();
    String name = segments.get(segments.size() - 1);

    return name;
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/okhttp/OkHttpResponseExtractor.kt
package com.baulsupp.oksocial.okhttp

import com.baulsupp.oksocial.output.ResponseExtractor
import com.baulsupp.oksocial.output.util.JsonUtil
import okhttp3.Response
import okio.BufferedSource

class OkHttpResponseExtractor : ResponseExtractor<Response> {
  override fun mimeType(response: Response): String? {
    val host = response.request().url().host()

    val mediaType = response.body()?.contentType()

    return when {
      host == "graph.facebook.com" && mediaType?.subtype() == "javascript" -> JsonUtil.JSON
      host == "dns.google.com" && mediaType?.subtype() == "x-javascript" -> JsonUtil.JSON
      else -> mediaType?.toString()
    }
  }

  override fun source(response: Response): BufferedSource = response.body()!!.source()

  override fun filename(response: Response): String {
    val segments = response.request().url().pathSegments()

    return segments[segments.size - 1]
  }
}
